k2/csrc/hash.h

```diff
@@ -0,0 +1,281 @@
+/**
+ * @brief
+ * hash
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#ifndef K2_CSRC_HASH_H_
+#define K2_CSRC_HASH_H_
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "k2/csrc/algorithms.h"
+#include "k2/csrc/array.h"
+#include "k2/csrc/context.h"
+#include "k2/csrc/eval.h"
+#include "k2/csrc/log.h"
+#include "k2/csrc/utils.h"
+
+namespace k2 {
+
+
+
+// __host__ __device__ version of CUDA's atomicCAS (copy and swap).  In the host
+// case it assumes the calling code is single-threaded and that `compare` is the
+// value that was just read from `address`, so it assumes it still has that
+// value and returns it.  The compiler should be able to optimize calling code.
+unsigned long long int __forceinline__ __host__ __device__ AtomicCAS(
+    unsigned long long int* address,
+    unsigned long long int compare,
+    unsigned long long int val) {
+#ifdef __CUDA_ARCH__
+  return atomicCAS(address, compare, val);
+#else
+  *address = val;
+  return compare;
+#endif
+}
+
+/*
+  How Hash32 works:
+    - It's a map from uint32_t to int32_t.  (All keys except
+      ~0 == -1 == UINT32_MAX are allowed).
+    - Each bucket has a struct { uint32_t key; int32_t value; }
+    - If there is no value present, we set the key to ~0 == -1.
+    - The number of buckets is a power of 2 provided by the user.
+    - When accessing hash[key], we use bucket_index == key % num_buckets,
+      leftover_index = 1 | ((key * 2) / num_buckets).  This is
```

Comment:
The documentation does not match the implementation.

There is no `2` in the implementation.